src/main/java/com/code_intelligence/jazzer/mutation/mutator/lang/IntegralMutatorFactory.java

```diff
@@ -0,0 +1,279 @@
+          return mutateImpl(value, prng);
+        }
+
+        @Override
+        public Long read(DataInputStream in) throws IOException {
+          return in.readLong();
+        }
+
+        @Override
+        public void write(Long value, DataOutputStream out) throws IOException {
+          out.writeLong(value);
+        }
+      });
+    } else {
+      return Optional.empty();
+    }
+  }
+
+  // Based on
+  // https://github.com/google/fuzztest/blob/a663ded6c36f050fbdc634a8fc81d553068d71d7/fuzztest/internal/domain.h#L1447
+  // SPDX: Apache-2.0
+  // Copyright 2022 Google LLC
+  private static abstract class AbstractIntegralMutator<T extends Number>
+      extends SerializingMutator<T> {
+    private static final long RANDOM_WALK_RANGE = 5;
+    private final long minValue;
+    private final long maxValue;
+    private final int largestMutableBitNegative;
+    private final int largestMutableBitPositive;
+    private final long[] specialValues;
+
+    AbstractIntegralMutator(
+        AnnotatedType type, long defaultMinValueForType, long defaultMaxValueForType) {
+      long minValue = defaultMinValueForType;
+      long maxValue = defaultMaxValueForType;
+      // InRange is not repeatable, so the loop body will apply exactly once.
+      for (Annotation annotation : type.getAnnotations()) {
+        if (annotation instanceof InRange) {
+          InRange inRange = (InRange) annotation;
+          if (inRange.min() != Long.MIN_VALUE) {
+            require(inRange.min() >= defaultMinValueForType,
+                format("@InRange.min=%d is out of range: %s", inRange.min(), type.getType()));
+            minValue = inRange.min();
+          }
+          if (inRange.max() != Long.MIN_VALUE) {
+            require(inRange.max() <= defaultMaxValueForType,
+                format("@InRange.max=%d is out of range: %s", inRange.max(), type.getType()));
+            maxValue = inRange.max();
+          }
+        }
+      }
+
+      require(minValue <= maxValue,
+          format("[%d, %d] is not a valid interval: %s", minValue, maxValue, type));
+      require(minValue != maxValue,
+          format(
+              "[%d, %d] can not be mutated, use a constant instead: %s", minValue, maxValue, type));
+      this.minValue = minValue;
+      this.maxValue = maxValue;
+      if (minValue >= 0) {
+        largestMutableBitNegative = 0;
+        largestMutableBitPositive = bitWidth(minValue ^ maxValue);
+      } else if (maxValue < 0) {
+        largestMutableBitNegative = bitWidth(minValue ^ maxValue);
+        largestMutableBitPositive = 0;
+      } else /* minValue < 0 && maxValue >= 0 */ {
+        largestMutableBitNegative = bitWidth(~minValue);
+        largestMutableBitPositive = bitWidth(maxValue);
+      }
+      this.specialValues = collectSpecialValues(minValue, maxValue);
+    }
+
+    private static long[] collectSpecialValues(long minValue, long maxValue) {
+      return LongStream.of(0, 1, minValue, maxValue)
```

Comment:
Why is all the `LongStream` code needed? This should return an array with 4 elements, shouldn't it?

Reply:
The stream chain has a `distinct`: On an `@InRange(min = 0) int`, we don't want to generate `0` twice as often as `1`.

Reply:
Yes, good point.